kernel/src/hil/flash.rs

```diff
@@ -4,141 +4,150 @@
+//!     fn erase_complete(&self, ret: Result<(), ReturnCode>) {}
 //! }
 //! ```
 
 use crate::returncode::ReturnCode;
 
-/// Flash errors returned in the callbacks.
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
-pub enum Error {
-    /// Success.
-    CommandComplete,
-
-    /// An error occurred during the flash operation.
-    FlashError,
-}
-
 pub trait HasClient<'a, C> {
     /// Set the client for this flash peripheral. The client will be called
     /// when operations complete.
     fn set_client(&'a self, client: &'a C);
 }
 
-/// A page of writable persistent flash memory.
-pub trait Flash {
-    /// Type of a single flash page for the given implementation.
-    type Page: AsMut<[u8]> + Default;
-
-    /// Read a page of flash into the buffer.
-    fn read_page(
+/// A page of writeable persistent flash memory.
+///
+/// `W`: Should be the minimum number of bytes that can be written
+///      in an operation.
+/// `E`: Should be the minimum number of bytes that can be erased
+///      in an operation.
+pub trait Flash<const W: usize, const E: usize> {
+    /// Read data from flash into a buffer.
+    ///
+    /// This function will read data stored in flash at `address` and
+    /// `length` into the buffer `buf`.
+    ///
+    /// Not all implementations support this, in that case `read_page`
+    /// should be used instead.
+    ///
+    /// On success returns nothing
+    /// On failure returns a `ReturnCode` and the buffer passed in.
+    fn read(
         &self,
-        page_number: usize,
-        buf: &'static mut Self::Page,
-    ) -> Result<(), (ReturnCode, &'static mut Self::Page)>;
+        address: usize,
+        length: usize,
+        buf: &'static mut [u8],
+    ) -> Result<(), (ReturnCode, &'static mut [u8])>;
 
-    /// Write a page of flash from the buffer.
-    fn write_page(
+    /// Write data from a buffer to flash.
+    ///
+    /// This function will write the buffer `buf` to the `address` specified
+    /// in flash.
+    ///
+    /// `address` should be aligned to `W`.
+    ///
+    /// This function will not erase the page first. The user of this function
+    /// must ensure that a page is erased before writing.
+    ///
+    /// On success returns nothing
+    /// On failure returns a `ReturnCode` and the buffer passed in.
+    fn write(
         &self,
-        page_number: usize,
-        buf: &'static mut Self::Page,
-    ) -> Result<(), (ReturnCode, &'static mut Self::Page)>;
+        address: usize,
+        buf: &'static mut [u8; W],
+    ) -> Result<(), (ReturnCode, &'static mut [u8; W])>;
 
-    /// Erase a page of flash by setting every byte to 0xFF.
-    fn erase_page(&self, page_number: usize) -> ReturnCode;
+    /// Erase a page of flash, setting every byte to 0xFF.
+    ///
+    /// On success returns nothing
+    /// On failure returns a `ReturnCode`.
+    fn erase_page(&self, page_number: usize) -> Result<(), ReturnCode>;
```

Comment:
I think it's awkward to have some interfaces use addresses and others use page numbers. Especially once this HIL is generic across page size, I think it would be valuable for the page/address translation to in some ergonomic way be captured in this interface (`get_page_from_address`?).

The easier change might be to have `erase_page` (`erase_pages`?) take and address and length, like the other interfaces, and simply `EINVAL` addr/len pairs that aren't clean pages.

Reply:
Yeah, I see what you mean. But specifying an address and length to erase seems strange as well. Flash pretty much only ever erases and entire page (while generally allows more fine grained read/write). Allowing the caller to specify an address just seems confusing. People would then expect the erase to erase that address and length, even when it can't.

Reply:
I wonder if all of these problems are ameliorated by the layered HIL approach. The "basic flash" HIL is all page-based, it exposes only `page_read`, `page_write`, `page_erase` and takes page indicies. It's generic across page size.

The "advanced flash" HIL is address-based, it exposes `read` and `write` which take address/length, and is generic across the minimum alignment.

Reply:
So the "basic flash" HIL will be what we currently have and there will be a more advanced one (which is this PR).

How do capsules work with the two though? Won't capsules just end up using one and we either loose functionality or don't support all hardware?